File: 0067-add-binary/0067-add-binary.java
class Solution {
    public String addBinary(String a, String b) {
        StringBuilder result = new StringBuilder();
        int i = a.length() - 1;  // Pointer to the last bit of a
        int j = b.length() - 1;  // Pointer to the last bit of b
        int carry = 0;           // Carry for the next bit

        // Process both strings from right to left
        while (i >= 0 || j >= 0 || carry != 0) {
            // Get the current bit of a (0 if out of bounds)
            int bitA = (i >= 0) ? a.charAt(i--) - '0' : 0;

            // Get the current bit of b (0 if out of bounds)
            int bitB = (j >= 0) ? b.charAt(j--) - '0' : 0;

            // Add the bits and the carry
            int sum = bitA + bitB + carry;

            // Determine the new carry and the current result bit
            carry = sum / 2;
            result.append(sum % 2);
        }

        // The result is in reverse order, so reverse it before returning
        return result.reverse().toString();
    }
}

File: 0067-add-binary/0067-add-binary.c
char* addBinary(const char* a, const char* b) {
    int i = strlen(a) - 1;
    int j = strlen(b) - 1;
    int carry = 0;

    // max possible length = max(lenA, lenB) + 2 (carry + '\0')
    int maxLen = (strlen(a) > strlen(b) ? strlen(a) : strlen(b)) + 2;
    char* result = (char*)malloc(maxLen);
    int k = 0;

    while (i >= 0 || j >= 0 || carry != 0) {
        int bitA = (i >= 0) ? a[i--] - '0' : 0;
        int bitB = (j >= 0) ? b[j--] - '0' : 0;

        int sum = bitA + bitB + carry;
        carry = sum / 2;

        result[k++] = (sum % 2) + '0';
    }

    result[k] = '\0';

    // reverse the string
    for (int l = 0, r = k - 1; l < r; l++, r--) {
        char tmp = result[l];
        result[l] = result[r];
        result[r] = tmp;
    }

    return result; // caller must free
}